src/Lumper.UI/Updater/Updater.cs

```diff
@@ -0,0 +1,420 @@
+                int read;
+                var length = (int)response.Content.Headers.ContentLength.Value;
+                var remaining = length;
+                while (!handler.Cancelled &&
+                       (read = await downloadStream.ReadAsync(
+                           buffer.AsMemory(0, int.Min(buffer.Length, remaining)),
+                           cts.Token)) >
+                       0)
+                {
+                    var prog = (float)read / length * 100;
+                    handler.UpdateProgress(prog, $"{float.Floor((1 - ((float)remaining / length)) * 100)}%");
+                    await stream.WriteAsync(buffer.AsMemory(0, read), cts.Token);
+                    remaining -= read;
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            if (ex is TaskCanceledException)
+                LogManager.GetCurrentClassLogger().Info("Download cancelled by user");
+            else
+                LogManager.GetCurrentClassLogger().Error(ex, "Failed to download!");
+
+            await stream.DisposeAsync();
+            return null;
+        }
+        finally
+        {
+            progressWindow?.Close();
+            stream.Close();
+        }
+
+        if (handler.Cancelled)
+            return null;
+
+        return stream;
+    }
+    /// <summary>
+    /// Returns the URL to the download link for the OS-specific version.
+    /// </summary>
+    /// <returns></returns>
+    private static string GetPath(GithubUpdate assets, OSPlatform OS)
+    {
+        var operatingSystem = "";
+
+        if (OS == OSPlatform.Windows)
+            operatingSystem = "win";
+        else if (OS == OSPlatform.Linux)
+            operatingSystem = "linux";
+        else
+            throw new Exception("Unsupported OS: " + nameof(OS));
+
+        foreach (Asset asset in assets.Assets)
+        {
+            if (asset.Name.Contains(operatingSystem, StringComparison.OrdinalIgnoreCase))
+            {
+                return asset.BrowserDownloadUrl;
+            }
+
+        }
+        throw new Exception("Could not find download link for " + nameof(OS));
+    }
+
+    /// <summary>
+    /// Downloads a file from the given URI and places it in fileName
+    /// </summary>
+    public static async Task DownloadFile(Uri uri, string fileName) => await HttpDownload(uri.AbsoluteUri, fileName);
+    /// <summary>
+    /// Downloads an update for the program, applies it, and then restarts itself with the new version
+    /// </summary>
+    /// <returns></returns>
+    public static async ValueTask Update()
+    {
+        GithubUpdate assets;
+        try
+        {
+            assets = await GetGithubUpdates();
+
+
+        }
+        catch (Exception ex)
+        {
+            ButtonResult result = await MessageBoxManager
+            .GetMessageBoxStandard(
+                "Error",
+                ex.Message, ButtonEnum.Ok)
+            .ShowWindowDialogAsync(Program.Desktop.MainWindow);
```

Comment:
We don't tend to use message boxes for showing errors since we have a built-in logger UI. The standard pattern we use is to create an NLog logger for any class that needs one with 

```cs
private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
```
then for erroring you'd use `Logger.Error(<your message>)`.